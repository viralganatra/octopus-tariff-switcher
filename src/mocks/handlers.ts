import { http, HttpResponse, graphql } from 'msw';
import {
  acceptTermsAndConditionsFxture,
  accountFixture,
  onboardingProcessFixture,
  productAgileFixture,
  productCosyFixture,
  productGoFixture,
  productsFixture,
  telemetryFixture,
  termsAndConditionsForProductFixture,
  unitRatesFixture,
  unitRates2020Fixture,
  telemetry2020Fixture,
} from './fixtures';

function extractYearFromDate(date: string) {
  return Number(date.split('-').at(0));
}

export const handlers = [
  http.get('https://api.octopus.energy/v1/products', () => {
    return HttpResponse.json(productsFixture);
  }),
  http.get('https://api.octopus.energy/v1/products/AGILE-24-10-01/', () => {
    return HttpResponse.json(productAgileFixture);
  }),
  http.get('https://api.octopus.energy/v1/products/COSY-22-12-08/', () => {
    return HttpResponse.json(productCosyFixture);
  }),
  http.get('https://api.octopus.energy/v1/products/GO-VAR-22-10-14/', () => {
    return HttpResponse.json(productGoFixture);
  }),
  http.get(
    'https://api.octopus.energy/v1/products/:tariffCode/electricity-tariffs/:productCode/standard-unit-rates',
    ({ request }) => {
      const url = new URL(request.url);
      const periodFrom = url.searchParams.get('period_from') as string;

      if (extractYearFromDate(periodFrom) === 2020) {
        return HttpResponse.json(unitRates2020Fixture);
      }

      return HttpResponse.json(unitRatesFixture);
    },
  ),
  http.post('https://api.mjml.io/v1/render', async ({ request }) => {
    const data = (await request.json()) as { mjml: string };

    return HttpResponse.json({
      html: data.mjml,
    });
  }),
  graphql.mutation('ObtainKrakenToken', () => {
    return HttpResponse.json({
      data: {
        obtainKrakenToken: {
          token: 'foo',
        },
      },
    });
  }),
  graphql.query('Account', () => {
    return HttpResponse.json({
      data: accountFixture,
    });
  }),
  graphql.query('smartMeterTelemetry', ({ variables }) => {
    if (extractYearFromDate(variables.start) === 2020) {
      return HttpResponse.json({
        data: telemetry2020Fixture,
      });
    }

    return HttpResponse.json({
      data: telemetryFixture,
    });
  }),
  graphql.query('TermsAndConditionsForProduct', () => {
    return HttpResponse.json({
      data: termsAndConditionsForProductFixture,
    });
  }),
  graphql.mutation('StartOnboardingProcess', () => {
    return HttpResponse.json({
      data: onboardingProcessFixture,
    });
  }),
  graphql.mutation('AcceptTermsAndConditions', () => {
    return HttpResponse.json({
      data: acceptTermsAndConditionsFxture,
    });
  }),
];
