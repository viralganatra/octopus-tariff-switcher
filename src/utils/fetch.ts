import { FetchError } from '../errors/fetch-error';

export async function getData(url: string) {
  const response = await fetch(url, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    throw new FetchError(`Request failed with status ${response.status}`);
  }

  return response.json();
}

export async function sendData({
  url,
  body,
  headers,
}: { url: string; body: Record<string, string>; headers?: Record<string, string> }) {
  const response = await fetch(url, {
    headers,
    method: 'POST',
    body: JSON.stringify(body),
  });

  if (!response.ok) {
    throw new FetchError(`Request failed with status ${response.status}`);
  }

  return response.json();
}
