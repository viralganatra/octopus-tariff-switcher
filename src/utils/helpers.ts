import { TZDateMini } from '@date-fns/tz';
import type { IsoDate, IsoDateTime, Url } from '../types/misc';

export function getDateFromApiIsoString(isoStringFromApi: string) {
  return new Date(isoStringFromApi);
}

export function getMsFromApiIsoString(isoStringFromApi: string) {
  return getDateFromApiIsoString(isoStringFromApi).getTime();
}

export function getDateInLocalTimeZone(date: Date) {
  return new TZDateMini(date).withTimeZone('Europe/London');
}

export function roundTo4Digits(num: number) {
  return Math.round(num * 1e4) / 1e4;
}

export function roundTo2Digits(num: number) {
  return Math.round(num * 1e2) / 1e2;
}

export function penceToPoundWithCurrency(num: number) {
  return `Â£${roundTo2Digits(num / 100).toFixed(2)}`;
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export function toIsoDateString(input: string) {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(input)) {
    throw new Error('Invalid ISO date format');
  }
  return input as IsoDate;
}

export function toIsoDateTime(input: string) {
  if (!/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/.test(input)) {
    throw new Error('Invalid ISO 8601 datetime format');
  }
  return input as IsoDateTime;
}

export function makeUrl(value: string): Url {
  try {
    new URL(value);
    return value as Url;
  } catch {
    throw new Error(`Invalid URL: ${value}`);
  }
}

export function scrubKeys<T>({ data, keysToScrub }: { data: T; keysToScrub: string[] }): T {
  const SCRUBBED_VALUE = '[SCRUBBED]';

  if (Array.isArray(data)) {
    return data.map((item) => scrubKeys({ data: item, keysToScrub })) as T;
  }

  if (data !== null && typeof data === 'object') {
    const result: Record<string, unknown> = {};

    for (const key of Object.keys(data)) {
      const value = (data as Record<string, unknown>)[key];

      if (keysToScrub.includes(key)) {
        result[key] = SCRUBBED_VALUE;
      } else {
        result[key] = scrubKeys({ data: value, keysToScrub });
      }
    }

    return result as T;
  }

  return data;
}
