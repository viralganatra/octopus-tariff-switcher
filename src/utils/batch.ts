import { sleep } from './helpers';

// Split an array of items into multiple batches,
// e.g. [1, 2, 3, 4, 5] -> [[1, 2], [3, 4], [5]]
export function chunkArray<T>(arr: T[], size: number): T[][] {
  return arr.length === 0 ? [] : [arr.slice(0, size), ...chunkArray(arr.slice(size), size)];
}

// Process batches of items, with a delay between each batch
export async function processBatches<T, R>({
  batches,
  processBatchItem,
  delayMs = 500,
}: {
  batches: T[][];
  processBatchItem: (item: T) => Promise<R>;
  delayMs?: number;
}) {
  const results: R[] = [];

  async function process(index: number) {
    if (index >= batches.length) {
      return;
    }

    const batch = batches[index];

    if (!batch) {
      throw new Error(`No batch at index ${index}`);
    }

    const batchResults = await Promise.all(batch.map(processBatchItem));

    results.push(...batchResults);

    if (index < batches.length - 1) {
      await sleep(delayMs);
      return process(index + 1);
    }
  }

  await process(0);

  return results;
}
