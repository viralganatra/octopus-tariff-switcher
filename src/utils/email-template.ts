import { penceToPoundWithCurrency } from './helpers';
import type { TariffContextWithCost } from '../types/tariff';

type TemplateData = {
  allTariffsByCost: TariffContextWithCost[];
  currentTariffWithCost: TariffContextWithCost;
  title: string;
};

type CommonTariffTemplate = {
  content: string;
  imageUrl: string;
} & Omit<TemplateData, 'currentTariffWithCost'>;

function wrapperEmailTemplate(content: string) {
  const htmlOutput = `
    <mjml>
      <mj-body background-color="white">
        <mj-section background-color="#f0f0f0">
          <mj-column width="15%" vertical-align="middle">
            <mj-image src="https://i.postimg.cc/v4fmRR64/businessman.png"></mj-image>
          </mj-column>
          <mj-column width="85%" vertical-align="middle">
        <mj-text font-style="bold" font-size="24px" color="#525252" font-family="Helvetica Neue">Octopus Tariff Switcher</mj-text>
       </mj-column>
      </mj-section>
        ${content}
      </mj-body>
    </mjml>
  `;

  return htmlOutput;
}

function getCommonTariffTemplate({
  allTariffsByCost,
  title,
  content,
  imageUrl,
}: CommonTariffTemplate) {
  const htmlOutput = `
    <mj-section>
      <mj-column width="60%">
        <mj-text font-style="bold" font-size="24px" line-height="30px" font-family="Helvetica Neue" color="#525252">${title}.</mj-text>
          <mj-table font-size="16px" line-height="24px" font-family="Helvetica Neue" color="#525252">
            <tr style="border-bottom:1px solid #ecedee;text-align:left;padding:15px 0;">
              <th style="padding: 8px">Tariff</th>
              <th style="padding: 8px">Cost</th>
            </tr>
            ${allTariffsByCost
              .map((tariff, index) => {
                const rowHighlight = index % 2 === 0 ? 'style="background-color:#f0f0f0"' : '';

                return `
                <tr ${rowHighlight}>
                  <td style="padding: 8px">${tariff.displayName}</td>
                  <td style="padding: 8px">${penceToPoundWithCurrency(tariff.cost)}</td>
                </tr>
              `;
              })
              .join('')}
          </mj-table>
          ${content}
        </mj-column>
        <mj-column width="40%">
          <mj-image src="${imageUrl}"></mj-image>
        </mj-column>
      </mj-section>
  `;

  return wrapperEmailTemplate(htmlOutput);
}

export function getCheaperTariffTemplate({
  allTariffsByCost,
  currentTariffWithCost,
  title,
}: TemplateData) {
  const cheapestTariff = allTariffsByCost.at(0) ?? currentTariffWithCost;
  const saving = penceToPoundWithCurrency(currentTariffWithCost.cost - cheapestTariff.cost);

  const htmlOutput = getCommonTariffTemplate({
    allTariffsByCost,
    title,
    content: `
      <mj-text font-size="16px" line-height="24px" padding-top="20px" color="#525252">
        You are currently on ${currentTariffWithCost.displayName}. I have automatically switched to ${cheapestTariff.displayName} as it is cheaper.
      </mj-text>
      <mj-text font-size="16px" line-height="24px" padding-top="10px" color="#525252">
        You will pay ${penceToPoundWithCurrency(cheapestTariff.cost)} versus ${penceToPoundWithCurrency(currentTariffWithCost.cost)}, saving ${saving}.
      </mj-text>
    `,
    imageUrl: 'https://i.postimg.cc/y8DtP8TQ/robot-cheaper-tariff.png',
  });

  return htmlOutput;
}

export function getAlreadyCheapestTariffTemplate({
  allTariffsByCost,
  currentTariffWithCost,
  title,
}: TemplateData) {
  const htmlOutput = getCommonTariffTemplate({
    allTariffsByCost,
    title,
    content: `
      <mj-text font-size="16px" line-height="24px" padding-top="20px" color="#525252">
        You are currently on ${currentTariffWithCost.displayName}. It isn't worth switching as the other tariffs are more expensive.
      </mj-text>
    `,
    imageUrl: 'https://i.postimg.cc/NfJZmHR2/robot-already-on-cheapest.png',
  });

  return htmlOutput;
}

export function notWorthSwitchingTariffTemplate({
  allTariffsByCost,
  currentTariffWithCost,
  title,
}: TemplateData) {
  const htmlOutput = getCommonTariffTemplate({
    allTariffsByCost,
    title,
    content: `
      <mj-text font-size="16px" line-height="24px" padding-top="20px" color="#525252">
        You are currently on ${currentTariffWithCost.displayName}. It isn't worth switching to the other tariffs as the saving is too little.
      </mj-text>
    `,
    imageUrl: 'https://i.postimg.cc/JzSgdC61/robot-not-worth-sw-itching.png',
  });

  return htmlOutput;
}
