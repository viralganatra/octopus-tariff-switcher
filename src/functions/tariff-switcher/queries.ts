import { Graffle } from 'graffle';
import { Resource } from 'sst';
import { z } from 'zod';
import { API_GRAPHQL, API_PRODUCTS } from '../../constants/api';
import { getData } from '../../utils/fetch';
import { formatISO } from 'date-fns';
import type { TariffSelector } from '../../types/tariff';

export async function fetchToken() {
  const schema = z.object({
    obtainKrakenToken: z.object({
      token: z.string(),
    }),
  });

  const graffle = Graffle.create().transport({
    url: API_GRAPHQL,
  });

  const result = await graffle.gql`
    mutation ObtainKrakenToken($input: ObtainJSONWebTokenInput!) {
      obtainKrakenToken(input: $input) {
        token
      }
    }
  `.send({ input: { APIKey: Resource.ApiKey.value } });

  const data = schema.parse(result);

  return data;
}

export async function fetchAccountInfo({ token }: { token: string }) {
  const schema = z.object({
    account: z.object({
      electricityAgreements: z
        .array(
          z.object({
            validFrom: z.string().datetime({ offset: true }),
            validTo: z.string().datetime({ offset: true }).nullable(),
            meterPoint: z.object({
              meters: z
                .array(
                  z.object({
                    smartDevices: z
                      .array(
                        z.object({
                          deviceId: z.string(),
                        }),
                      )
                      .nonempty(),
                  }),
                )
                .nonempty(),
            }),
            tariff: z.object({
              tariffCode: z.string(),
              standingCharge: z.number(),
            }),
          }),
        )
        .nonempty(),
    }),
  });

  const graffle = Graffle.create().transport({
    url: API_GRAPHQL,
    headers: {
      authorization: token,
    },
  });

  const result = await graffle.gql`
      query Account($accountNumber: String!) {
        account(accountNumber: $accountNumber) {
          electricityAgreements(active: true) {
            validFrom
            validTo
            meterPoint {
              meters(includeInactive: false) {
                smartDevices {
                  deviceId
                }
              }
              mpan
            }
            tariff {
              ... on HalfHourlyTariff {
                productCode
                tariffCode
                standingCharge
              }
            }
          }
        }
      }
    `.send({ accountNumber: Resource.AccNumber.value });

  const results = schema.parse(result);

  return results;
}

export async function fetchSmartMeterTelemetry({
  token,
  deviceId,
  startDate,
  endDate,
}: { token: string; deviceId: string; startDate: string; endDate: string }) {
  const schema = z.object({
    smartMeterTelemetry: z
      .array(
        z.object({
          readAt: z.string(),
          consumptionDelta: z.coerce.number(),
          costDeltaWithTax: z.coerce.number(),
        }),
      )
      .nonempty(),
  });

  const graffle = Graffle.create().transport({
    url: API_GRAPHQL,
    headers: {
      authorization: token,
    },
  });

  // consumptionDelta - Energy consumption in Wh between the read_at and the next reading.
  // costDeltaWithTax - Energy cost including VAT for the consumption delta in pence.
  // readAt - The start_at time of the telemetry data
  const result = await graffle.gql`
    query smartMeterTelemetry(
      $deviceId: String!,
      $start: DateTime,
      $end: DateTime,
      $grouping: TelemetryGrouping
    ) {
      smartMeterTelemetry(
        deviceId: $deviceId,
        start: $start,
        end: $end,
        grouping: $grouping
      ) {
        readAt
        consumptionDelta
        costDeltaWithTax
      }
    }
  `.send({ deviceId, start: startDate, end: endDate, grouping: 'HALF_HOURLY' });

  const results = schema.parse(result);

  return results;
}

export async function fetchAllProducts() {
  const schema = z.object({
    results: z.array(
      z.object({
        display_name: z.string(),
        direction: z.enum(['IMPORT', 'EXPORT']),
        brand: z.string(),
        code: z.string(),
      }),
    ),
  });

  const data = await getData(`${API_PRODUCTS}?brand=OCTOPUS_ENERGY&is_business=false`);

  const { results } = schema.parse(data);

  return results;
}

export async function fetchTodaysUnitRatesByTariff({ tariffCode, productCode }: TariffSelector) {
  const schema = z.object({
    results: z
      .array(
        z.object({
          value_inc_vat: z.number(),
          valid_from: z.string().datetime(),
          valid_to: z.string().datetime(),
        }),
      )
      .nonempty(),
  });

  const today = formatISO(new Date(), { representation: 'date' });

  const data = await getData(
    `${API_PRODUCTS}/${tariffCode}/electricity-tariffs/${productCode}/standard-unit-rates/?period_from=${today}T00:00:00Z&period_to=${today}T23:59:59Z`,
  );

  const { results } = schema.parse(data);

  return results;
}

export async function fetchTodaysStandingCharge({ tariffCode, productCode }: TariffSelector) {
  const schema = z.object({
    results: z
      .array(
        z.object({
          value_inc_vat: z.number(),
        }),
      )
      .nonempty(),
  });

  const data = await getData(
    `${API_PRODUCTS}/${tariffCode}/electricity-tariffs/${productCode}/standing-charges/`,
  );

  const { results } = schema.parse(data);

  return results;
}
