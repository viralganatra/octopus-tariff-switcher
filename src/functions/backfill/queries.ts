import { Resource } from 'sst';
import { API, API_ACCOUNTS, API_PRODUCTS } from '../../constants/api';
import { logger } from '../../utils/logger';
import { getData } from '../../utils/fetch';
import { makeUrl, scrubKeys, roundTo4Digits } from '../../utils/helpers';
import type { TariffSelectorWithDate } from '../../types/tariff';
import { schemaConsumption, schemaPastTariffs, schemaStandingCharge } from './schema';
import { fetchToken } from '../tariff-switcher/queries';
import type { IsoDate } from '../../types/misc';

export async function fetchAllPastTariffs() {
  const token = await fetchToken();

  logger.info('API: Getting past tariffs');

  const result = await getData({
    url: makeUrl(`${API_ACCOUNTS}/${Resource.AccNumber.value}/`),
    headers: {
      authorization: token,
    },
  });

  const scrubbedResult = scrubKeys({
    data: result,
    keysToScrub: [
      'moved_in_at',
      'moved_out_at',
      'address_line_1',
      'address_line_2',
      'address_line_3',
      'town',
      'county',
      'postcode',
    ],
  });

  logger.info('API Response: Recieved past tariffs', { apiResponse: scrubbedResult });

  const results = schemaPastTariffs.parse(scrubbedResult);

  const { agreements } = results.properties[0].electricityMeterPoints[0];

  return agreements;
}

export async function fetchConsumption({
  mpan,
  serialNumber,
  isoDate,
}: { mpan: string; serialNumber: string; isoDate: IsoDate }) {
  const token = await fetchToken();

  const params = {
    period_from: `${isoDate}T00:00:00Z`,
    period_to: `${isoDate}T23:59:59Z`,
    order_by: 'period',
  };

  const queryParams = new URLSearchParams(params).toString();

  const url = makeUrl(
    `${API}electricity-meter-points/${mpan}/meters/${serialNumber}/consumption/?${queryParams}`,
  );

  logger.info('API: Getting consumption', {
    data: { url, date: isoDate },
  });

  const result = await getData({
    url,
    headers: {
      authorization: token,
    },
  });

  logger.info('API Response: Recieved consumption', {
    date: isoDate,
    apiResponse: result,
  });

  const { results } = schemaConsumption.parse(result);

  return results;
}

export async function fetchStandingCharge({
  productCode,
  tariffCode,
  isoDate,
}: TariffSelectorWithDate) {
  const token = await fetchToken();

  const params = {
    period_from: `${isoDate}T00:00:00Z`,
    period_to: `${isoDate}T23:59:59Z`,
  };

  const queryParams = new URLSearchParams(params).toString();

  const url = makeUrl(
    `${API_PRODUCTS}/${productCode}/electricity-tariffs/${tariffCode}/standing-charges/?${queryParams}`,
  );

  logger.info('API: Getting standing charge', {
    data: { productCode, tariffCode, isoDate },
  });

  const result = await getData({
    url,
    headers: {
      authorization: token,
    },
  });

  logger.info('API Response: Recieved standing charge', {
    data: { productCode, tariffCode, isoDate },
    apiResponse: result,
  });

  const { results } = schemaStandingCharge.parse(result);

  return roundTo4Digits(results[0].valueIncVat);
}
