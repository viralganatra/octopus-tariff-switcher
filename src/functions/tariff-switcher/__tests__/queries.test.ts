import { server } from '../../../mocks/node';
import { toIsoDateString } from '../../../utils/helpers';
import { fetchAllProducts, fetchToken, fetchUnitRatesByTariff } from '../queries';

describe('Queries', () => {
  it('should return a token from the api or reuse it if it exists', async () => {
    const dispatchRequest = vi.fn();
    server.events.on('request:start', dispatchRequest);

    const token = await fetchToken();
    await fetchToken();
    await fetchToken();

    expect(token).toBe('foo');
    expect(dispatchRequest).toHaveBeenCalledOnce();

    const serverRequest = dispatchRequest.mock.lastCall?.at(0).request;

    await expect(serverRequest.json()).resolves.toMatchInlineSnapshot(`
      {
        "query": "
          mutation ObtainKrakenToken($input: ObtainJSONWebTokenInput!) {
            obtainKrakenToken(input: $input) {
              token
            }
          }
        ",
        "variables": {
          "input": {
            "APIKey": "API_KEY",
          },
        },
      }
    `);

    expect(serverRequest.url).toBe('https://api.octopus.energy/v1/graphql/');
  });

  it('should use the date param or today when fetching unit rates by tariff', async () => {
    await expect(
      fetchUnitRatesByTariff({
        tariffCode: 'E-1R-AGILE-18-02-21-A',
        productCode: 'AGILE-18-02-21',
      }),
    ).resolves.toMatchSnapshot();

    await expect(
      fetchUnitRatesByTariff({
        tariffCode: 'E-1R-AGILE-18-02-21-A',
        productCode: 'AGILE-18-02-21',
        isoDate: toIsoDateString('2020-01-01'),
      }),
    ).resolves.toMatchSnapshot();
  });

  it('should only fetch the list of all products once', async () => {
    const dispatchRequest = vi.fn();
    server.events.on('request:start', dispatchRequest);

    await fetchAllProducts();
    await fetchAllProducts();
    await fetchAllProducts();

    expect(dispatchRequest).toHaveBeenCalledOnce();
  });
});
