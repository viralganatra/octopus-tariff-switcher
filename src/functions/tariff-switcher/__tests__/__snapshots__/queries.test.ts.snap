// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Queries > should use the date param or today when fetching unit rates by tariff 1`] = `
[
  {
    "valid_from": "2025-03-03T22:00:00Z",
    "valid_to": "2025-03-02T00:00:00Z",
    "value_inc_vat": 13.23168,
  },
  {
    "valid_from": "2025-03-03T19:00:00Z",
    "valid_to": "2025-03-03T22:00:00Z",
    "value_inc_vat": 26.98143,
  },
  {
    "valid_from": "2025-03-03T16:00:00Z",
    "valid_to": "2025-03-03T19:00:00Z",
    "value_inc_vat": 40.47225,
  },
  {
    "valid_from": "2025-03-03T13:00:00Z",
    "valid_to": "2025-03-03T16:00:00Z",
    "value_inc_vat": 13.23168,
  },
  {
    "valid_from": "2025-03-03T07:00:00Z",
    "valid_to": "2025-03-03T13:00:00Z",
    "value_inc_vat": 26.98143,
  },
  {
    "valid_from": "2025-03-03T04:00:00Z",
    "valid_to": "2025-03-03T07:00:00Z",
    "value_inc_vat": 13.23168,
  },
  {
    "valid_from": "2025-03-03T00:00:00Z",
    "valid_to": "2025-03-03T04:00:00Z",
    "value_inc_vat": 26.98143,
  },
]
`;

exports[`Queries > should use the date param or today when fetching unit rates by tariff 2`] = `
[
  {
    "valid_from": "2020-03-03T22:00:00Z",
    "valid_to": "2020-03-02T00:00:00Z",
    "value_inc_vat": 13.23168,
  },
]
`;
